codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java

```diff
@@ -0,0 +1,266 @@
+
+    private final Http2Connection connection;
+    private final Http2Connection.PropertyKey stateKey;
+    private final WriteVisitor writeVisitor = new WriteVisitor();
+    private final Deque<State> queue = new ArrayDeque<State>();
+    private final EnqueueStrategy defaultEnqueueStrategy = new EnqueueStrategy();
+
+    /**
+     * The first time a state is enqueued, it will be added to the head of the queue. After
+     * that it will be added to the tail.
+     */
+    private final EnqueueStrategy firstEnqueueStrategy = new EnqueueStrategy() {
+        @Override
+        void enqueue(State state) {
+            queue.addFirst(state);
+            state.enqueued = true;
+
+            // Switch to the default strategy so that it will be added to the tail from
+            // now on.
+            state.enqueueStrategy = defaultEnqueueStrategy;
+        }
+    };
+
+    /**
+     * The minimum number of bytes that we will attempt to allocate to a stream. This is to
+     * help improve goodput on a per-stream basis.
+     */
+    private final int minAllocationChunk;
+    private long totalStreamableBytes;
+
+    public UniformStreamByteDistributor(Http2Connection connection) {
+        this(connection, DEFAULT_MIN_ALLOCATION_CHUNK);
+    }
+
+    public UniformStreamByteDistributor(Http2Connection connection, int minAllocationChunk) {
+        this.connection = checkNotNull(connection, "connection");
+        if (minAllocationChunk <= 0) {
+            throw new IllegalArgumentException("minAllocationChunk must be > 0");
+        }
+        this.minAllocationChunk = minAllocationChunk;
+
+        // Add a state for the connection.
+        stateKey = connection.newKey();
+        Http2Stream connectionStream = connection.connectionStream();
+        connectionStream.setProperty(stateKey, new State());
+
+        // Register for notification of new streams.
+        connection.addListener(new Http2ConnectionAdapter() {
+            @Override
+            public void onStreamAdded(Http2Stream stream) {
+                stream.setProperty(stateKey, new State());
+            }
+
+            @Override
+            public void onStreamClosed(Http2Stream stream) {
+                state(stream).close();
+            }
+        });
+    }
+
+    @Override
+    public void updateStreamableBytes(StreamState streamState) {
+        State state = state(streamState.stream());
+        state.updateStreamableBytes(streamState.streamableBytes(), streamState.hasFrame());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
+        checkNotNull(writer, "writer");
+
+        // Allocate the bytes to the streams.
+        long allocated = allocateBytes(maxBytes);
+
+        long unallocatedStreamableBytes = totalStreamableBytes - allocated;
+
+        // Need to write even if maxBytes == 0 in order to handle the case of empty frames.
+        writeVisitor.writeAllocatedBytes(writer);
```

Comment:
If you have 100 idle streams, I would expect a noticeable difference. Not having to loop over all streams was a big reason I was wanting to change the algorithm. Any scheduling algorithm that has to loop over all items when choosing the next "is doing it wrong."

It is true that we can reduce number of data frames by write combining. However, the gains are expected to be small. If your `minAllocationChunk` is 1K, then the 9 byte frame overhead is < 1%. If the `minAllocationChunk` is 16K then the frame overhead is ~0.05%. Yes, you can have lower overhead by doing more combining, but it simply stops mattering at a point. Also, the cases you want a small `minAllocationChunk` are also the cases you don't want to combine the writes, because the link is slow and you don't want to monopolize the connection with a single stream as long.

@Scottmitch, I'm okay with committing and iterating.